presto-main/src/main/java/io/prestosql/type/TimeWithTimeZoneOperators.java

```diff
@@ -106,12 +117,23 @@ public static boolean greaterThanOrEqual(@SqlType(StandardTypes.TIME_WITH_TIME_Z
     }
 
     @ScalarOperator(CAST)
-    @SqlType(StandardTypes.TIME)
-    public static long castToTime(ConnectorSession session, @SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long value)
+    @LiteralParameters("p")
+    @SqlType("time(p)")
+    public static long castToTime(@LiteralParameter("p") long precision, @SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long value)
     {
-        // This is exactly the same operation as for TIME WITH TIME ZONE -> TIMESTAMP, as the representations
-        // of those types are aligned in range that is covered by TIME WITH TIME ZONE.
-        return TimeWithTimezoneToTimestampCast.cast(3, session, value);
+        long millis = unpackMillisUtc(value);
+        if (millis < 0) {
+            millis += MILLISECONDS_PER_DAY;
+        }
+        LocalDateTime localDateTime = LocalDateTime.of(LocalDate.ofEpochDay(0), LocalTime.ofNanoOfDay(millis / PICOSECONDS_PER_NANOSECOND));
```

Comment:
That looks wrong. The `millis` should be interpreted in the zone packed in `value`:

```
LocalTime.ofInstant(Instant.ofEpochMilli(millis), unpackZoneKey(value).getZoneId())
```